src/lib/node_status_service/node_status_service.ml

```diff
@@ -0,0 +1,396 @@
+  ; get_transition_knowledge : int
+  ; get_transition_chain_proof : int
+  ; get_node_status : int
+  ; get_ancestry : int
+  ; ban_notify : int
+  ; get_best_tip : int
+  ; get_epoch_ledger : int
+  }
+[@@deriving to_yojson]
+
+type gossip_count =
+  { new_state : int; transaction_pool_diff : int; snark_pool_diff : int }
+[@@deriving to_yojson]
+
+type block =
+  { hash : Mina_base.State_hash.t
+  ; sender : Network_peer.Envelope.Sender.t
+  ; received_at : string
+  ; is_valid : bool
+  ; reason_for_rejection :
+      [ `Invalid_proof
+      | `Invalid_delta_transition_chain_proof
+      | `Too_early
+      | `Too_late
+      | `Invalid_genesis_protocol_state
+      | `Invalid_protocol_version
+      | `Mismatched_protocol_version ]
+      option
+  }
+[@@deriving to_yojson]
+
+type node_status_data =
+  { block_height_at_best_tip : int
+  ; max_observed_block_height : int
+  ; max_observed_unvalidated_block_height : int
+  ; catchup_job_states : catchup_job_states option
+  ; sync_status : Sync_status.t
+  ; libp2p_input_bandwidth : float
+  ; libp2p_output_bandwidth : float
+  ; libp2p_cpu_usage : float
+  ; commit_hash : string
+  ; git_branch : string
+  ; peer_id : string
+  ; ip_address : string
+  ; timestamp : string
+  ; uptime_of_node : float
+  ; peer_count : int
+  ; rpc_received : rpc_count
+  ; rpc_sent : rpc_count
+  ; pubsub_msg_received : gossip_count
+  ; pubsub_msg_broadcasted : gossip_count
+  ; received_blocks : block list
+  }
+[@@deriving to_yojson]
+
+let send_node_status_data ~logger ~url node_status_data =
+  let node_status_json = node_status_data_to_yojson node_status_data in
+  let json = `Assoc [ ("data", node_status_json) ] in
+  let headers =
+    Cohttp.Header.of_list [ ("Content-Type", "application/json") ]
+  in
+  match%map
+    Async.try_with (fun () ->
+        Cohttp_async.Client.post ~headers
+          ~body:(Yojson.Safe.to_string json |> Cohttp_async.Body.of_string)
+          url)
+  with
+  | Ok ({ status; _ }, body) ->
+      let metadata =
+        [ ("data", node_status_json); ("url", `String (Uri.to_string url)) ]
+      in
+      if Cohttp.Code.code_of_status status = 200 then
+        [%log info] "Sent node status data to URL $url" ~metadata
+      else
+        let extra_metadata =
+          match body with
+          | `String s ->
+              [ ("error", `String s) ]
+          | `Strings ss ->
+              [ ("error", `List (List.map ss ~f:(fun s -> `String s))) ]
+          | `Empty | `Pipe _ ->
+              []
+        in
+        [%log error] "Failed to send node status data to URL $url"
+          ~metadata:(metadata @ extra_metadata)
+  | Error e ->
+      [%log error] "Failed to send node status data to URL $url"
+        ~metadata:
+          [ ("error", `String (Exn.to_string e))
+          ; ("url", `String (Uri.to_string url))
+          ]
+
+let reset_gauges () =
+  Mina_metrics.(
+    Gauge.set (snd Network.get_some_initial_peers_rpcs_sent) 0. ;
```

Comment:
could you make this into a loop, instead of repeating `Gauge.set (snd M) 0`, for each module `M`?